Electrum support

ysangkok opened the pull request:
This adds Electrum into the framework, it depends on ElectrumX and Electrum being checked out. I didn't add them into the Makefile because the setup procedure is documented on our repository and probably changes too often to be conveniently committed to here.

Please let me know if there is anything you'd like me to fix. This PR includes the other pull requests that are currently pending.

Eclair currently doesn't work for reasons documented in the other PR (Eclair doesn't support separate block/tx ports for ZMQ, while LND demands it). But the issue will presumably be fixed in Eclair. It is unrelated to the Electrum support. If you want to merge the other PR's separately, I can remove the Eclair and LND stuff from here.

ysangkok committed:
Add Electrum

cdecker commented on lightningd.py:
```diff
@@ -55,7 +55,7 @@ class LightningNode(object):
     displayName = 'lightning'
 
-    def __init__(self, lightning_dir, lightning_port, btc, executor=None, node_id=0):
+    def __init__(self, lightning_dir, lightning_port, btc, electrumx, executor=None, node_id=0):
```
Why add the dependency on `electrumx` if `lightningd` can run perfectly fine without it?

cdecker commented on eclair.py:
```diff
@@ -46,7 +46,7 @@ def __init__(self, lightning_dir, bitcoin_dir, port):
 
         self.cmd_line = [
-            '/usr/lib/jvm/java-8-openjdk-amd64/bin/java',
+            'java',
```
With PR #39 merged this commit is now redundant

cdecker commented on eclair.py:
```diff
@@ -114,7 +114,7 @@ class EclairNode(object):
     displayName = 'eclair'
 
-    def __init__(self, lightning_dir, lightning_port, btc, executor=None,
+    def __init__(self, lightning_dir, lightning_port, btc, electrumx, executor=None,
```
Notice that fixtures are instantiated on demand, so adding it as an argument here will start an `electrumx` instance on every test involving `eclair`. Since `eclair` doesn't use electrum this is not necessary and will slow down testing.

cdecker commented on lnd.py:
```diff
@@ -78,7 +78,7 @@ class LndNode(object):
     displayName = 'lnd'
 
-    def __init__(self, lightning_dir, lightning_port, btc, executor=None, node_id=0):
+    def __init__(self, lightning_dir, lightning_port, btc, electrumx, executor=None, node_id=0):
```
Same as with `eclair` and `lightningd`

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+
+class ElectrumX:
+    def __init__(self):
+        print("server/__init__")
```
Please use the logging framework for messages, otherwise we can't control the output and it'll be hard to tune to find issues.

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+        print("server/start")
+        from os import environ
+        rpc_port = 8000
+        os.system("rm -rf COIN hist meta utxo")
```
Don't use `os.system` and don't use `rm -rf` in there.

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+
+    def start(self):
+        print("server/start")
+        from os import environ
```
Don't use inline imports.

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+from electrum import simple_config, constants, keystore, util
+from electrum.wallet import Wallet
+from electrum.storage import WalletStorage
+from electrum.mnemonic import Mnemonic
```
Please sort imports.

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+import os
+import tempfile
+import traceback
+import asyncio
```
I'm rather hesitant to mix concurrency paradigms, and the rest uses the `threading` module. Would you mind using that for the time being?

ysangkok replied:
Since both Electrum and ElectrumX rely on Asyncio, and we need to somehow schedule Tasks on the loop of Electrum, and somehow launch and shutdown ElectrumX (which also uses asyncio), I don't see how I can avoid importing Asyncio here.

1. One way of doing this without importing Asyncio in lightning-integration would entail having the call to run_coroutine_threadsafe in the Electrum codebase proper, and then simply calling that method from the lightning-integration test, but since it would be a method only used here, I don't see the value of having that in Electrum. By the way, we are already mixing concurrency paradigms in Electrum anyway, since e.g. Qt doesn't natively support the Asyncio loop. It is a necessary evil since not all software is written using the same concurrency paradigms. 
1. Another way to do it, would be to parse the log of some process, but I don't like this approach of string based RPC since the shared implementation language allows lightning-integration's Electrum parts to easily drill into Electrum structures. Why would we settle on some logging format and then parse it here?
1. Another way would be to make all of lightning-integration use asyncio (good idea but I don't have the time), or rolling back Electrum to use plain old threads, which makes no sense, Asyncio is working great for us.

Of all these alternatives, the most attractive one the first one, and I could do that, but I'd really prefer not to, because it would just be another layer of pointless indirection, since it'd be an abstraction that is exclusively used here.

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+            "SSL_PORT": "51002",
+            "RPC_PORT": str(8000),
+            "NET": "regtest",
+            "DAEMON_URL": "http://rpcuser:rpcpass@127.0.0.1:" + str(BITCOIND_CONFIG.get('rpcport', 18332)),
```
I recently broke this, by making all ports variable in order to enable parallel test execution, sorry about that. Could you use the `ephemeral_port_reserve.reserve` function to self-assign a unique port?

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+
+        self.thread = threading.Thread(target=target)
+        self.thread.start()
+        #controller.run()
```
Please remove commented code.

cdecker commented on electrumutils.py:
```diff
@@ -0,0 +1,212 @@
+            "server": "localhost:51001:t",
+            "request_initial_sync": False,
+        }
+        user_dir = tempfile.mkdtemp(prefix="lightning-integration-electrum-")
```
The fixtures now contain a fixture called directory which is just a path. That directory is test-specific and should be used as the root for all things related to that test. Could you use that instead, so we can clean up tests that pass correctly?

cdecker commented on ptarmd.py:
```diff
@@ -44,7 +44,7 @@ class PtarmNode(object):
     displayName = 'ptarmigan'
 
-    def __init__(self, lightning_dir, lightning_port, btc, executor=None, node_id=0):
+    def __init__(self, lightning_dir, lightning_port, btc, electrumx, executor=None, node_id=0):
```
And once more :wink:

cdecker commented on test.py:
```diff
@@ -30,13 +31,14 @@
     """A factory to setup and start `lightningd` daemons.
     """
-    def __init__(self, testname, executor, btc, btcd):
+    def __init__(self, testname, executor, btc, btcd, electrumx):
```
Oh I see, we're not really injecting fixtures here, we're instantiating a `NodeFactory`. Let's see if we can make this a bit more lightweight somehow.

cdecker replied:
Maybe having a reference to the `NodeFactory` and pulling in the `electrumx` dependency on demand might be the key here.

cdecker commented on electrumutils.py:
```diff
@@ -37,8 +37,8 @@ def start(self):
         environ.update({
             "COIN": "BitcoinSegwit",
-            "SSL_KEYFILE": "/home/janus/electrumx/key.pem",
-            "SSL_CERTFILE": "/home/janus/electrumx/cert.pem",
```
Should be squashed since otherwise this breaks bisectability.

cdecker commented on electrumutils.py:
```diff
@@ -97,7 +97,7 @@ def start(self):
         wallet.start_network(self.actual.network)
         self.actual.add_wallet(wallet)
-        self.actual.start()
+        #self.actual.start()
```
Remove commented code please.

cdecker commented on lnd.py:
```diff
@@ -13,13 +13,11 @@
 
 # Needed for grpc to negotiate a valid cipher suite
-os.environ["GRPC_SSL_CIPHER_SUITES"] = "ECDHE-ECDSA-AES256-GCM-SHA384"
+os.environ["GRPC_SSL_CIPHER_SUITES"] = "HIGH+ECDSA"
```
Commit is redundant with #38, please mark as such.

cdecker commented on electrumutils.py:
```diff
@@ -125,7 +125,7 @@ def id(self):
         print("node/openchannel")
         # second parameter is local_amt_sat not capacity!
-        r = self.wallet.lnworker.open_channel(node_id, satoshis, 0, pw=None)
+        r = self.wallet.lnworker.open_channel(node_id, satoshis, 0)
```
Can also be squashed with the first commit.

cdecker commented on test.py:
```diff
@@ -115,8 +115,9 @@ def btcd():
-def electrumx():
+@pytest.fixture()
+def electrumx(bitcoind):
+    # bitcoind only passed to ensure dependency topology (order)
```
Good catch :+1:

cdecker reviewed (changes requested):
Thanks @ysangkok, excellent PR, and sorry for the delayed feedback, I must've missed this.

I was rather pedantic in the review, but I thought I might start enforcing some code quality at some point, and yes I know my own code is not compliant with some of the criticism either :wink:

cdecker commented:
Regarding the existence of `electrumx` on the system, could you amend the `Dockerfile` in order to install that from a trustworthy source? That would ensure that we can run this on our nightly test system :wink:

ysangkok committed:
Read bytes for passing to ssl_channel_arguments

As documented on https://grpc.io/grpc/python/grpc.html#grpc.ssl_channel_credentials, this should be bytes. Tests fail on Python 3 if I don't do this.

ysangkok commented:
OK, I believe I addressed all your concerns except the mixing of concurrency paradigms. Awaiting your comments.

cdecker commented:
Oh, I didn't realize that you are starting the electrum daemon without making it a subprocess, instead running it as a separate thread.

That breaks isolation between tests and doesn't really match the way we do things for other daemons. I'm sure electrum has a standalone binary that can be started and controlled from the command line, could you please use that, and checkout the source code of electrum into `src/electrum` so we have more control over the version being tested.

dmkxyz commented:
@cdecker Trying to figure out this long-standing PR. It looks like the author tried to install Electrumx from the DockerFile and import/run that installation in test.py. 
What I understand we want is to instead clone and setup electrum into src/electrum from Makefile and use a binary from there, just like other implementations. 
Is that a correct understanding? If yes, I can try patching this PR with that change.

cdecker commented:
Hi @deepakkathayat thanks for picking this one up again. The issue is that I'd like to be able to manage the electrum version separately from the docker version (all other implementations always build `master`, since testing only "released" versions doesn't really give us any new information about compatibility).

In addition being able to run electrum in a separate process, using just the exposed interface also ensures that there is no lingering electrum instance in the tester process once we are done with the testcase.

dmkxyz commented:
@cdecker, understood. Is there an irc channel where one can discuss the project and this PR in detail?

cdecker commented:
We can probably go to #lightning-dev or #c-lightning

jtimon commented on test.py:
```diff
@@ -11,6 +11,7 @@
 from concurrent import futures
 from utils import BitcoinD, BtcD
 from bech32 import bech32_decode
+from electrumutils import ElectrumX, ElectrumNode
```
Seems like ElectrumX isn't needed here, no?

ysangkok replied:
@jtimon True

ysangkok commented:
@deepakkathayat I'll close this since I have left Electrum, but please feel free to use this.
